File: backend/app/api/agents.py
from __future__ import annotations

import json
import traceback
from collections.abc import Iterator
from typing import Literal

from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field
from sqlalchemy import desc, func

from app.agents.chapter_edit import LLMError, propose_chapter_edit
from app.agents.context import build_context_preview
from app.agents.workflows import analyze_chapter, check_consistency, draft_chapter, draft_chapter_stream, polish_chapter, analyze_volume
from app.agents_v3.persistence import WorkflowPersistence
from app.db.repository import rows_to_dicts
from app.db.session import get_business_db
from app.models.business import Chapter, ChapterSummary, GenerationLog, GenerationVersion
from app.schemas.models import ChapterAnalyzeRequest, ChapterDraftRequest, ConsistencyCheckRequest, PolishRequest, VolumeAnalyzeRequest


router = APIRouter()


class ChapterEditChatRequest(BaseModel):
    """章节对话改稿请求；正文由前端传入的编辑快照决定。"""

    project_id: int
    chapter_id: int
    chapter_no: int = 1
    chapter_title: str = ""
    outline_id: int | None = None
    content: str = Field(max_length=500_000)
    scope: Literal["chapter", "selection"] = "chapter"
    selection_start: int | None = None
    selection_end: int | None = None
    instruction: str = Field(min_length=1, max_length=5_000)
    conversation: list[dict[str, str]] = Field(default_factory=list)
    context_selection: dict[str, list[int]] | None = None


class ChapterEditApplyRequest(BaseModel):
    """确认候选改稿请求，带原稿快照用于并发冲突保护。"""

    project_id: int
    chapter_id: int
    expected_content: str = Field(max_length=500_000)
    revised_content: str = Field(min_length=1, max_length=500_000)
    summary: str = Field(default="", max_length=500)


@router.get("/{project_id}/logs")
def generation_logs(project_id: int, limit: int = 20) -> list[dict]:
    """读取 Agent 运行日志。

    前端工作台用它展示最近一次生成、分析、精修的轨迹；limit 做上限保护，
    避免日志很多时一次性传输过多内容。
    """
    safe_limit = max(1, min(limit, 100))
    with get_business_db() as db:
        logs = (
            db.query(GenerationLog)
            .filter(GenerationLog.project_id == project_id)
            .order_by(GenerationLog.id.desc())
            .limit(safe_limit)
            .all()
        )
    return rows_to_dicts(logs)


@router.get("/{project_id}/summaries")
def chapter_summaries(project_id: int, limit: int = 20) -> list[dict]:
    """读取章节长期记忆摘要。

    工作台右侧只需要最近摘要；后续若接入向量检索，可继续保持这个轻量列表入口。
    """
    safe_limit = max(1, min(limit, 100))
    with get_business_db() as db:
        rows = (
            db.query(
                ChapterSummary.id,
                ChapterSummary.chapter_id,
                Chapter.chapter_no,
                Chapter.title,
                ChapterSummary.summary,
                ChapterSummary.character_changes,
                ChapterSummary.world_changes,
                ChapterSummary.new_foreshadowings,
                ChapterSummary.timeline_events,
                ChapterSummary.source_run_id,
                ChapterSummary.source_version_id,
                GenerationVersion.version_number.label("source_version_number"),
                ChapterSummary.created_at,
            )
            .join(Chapter, Chapter.id == ChapterSummary.chapter_id)
            .outerjoin(GenerationVersion, GenerationVersion.version_id == ChapterSummary.source_version_id)
            .filter(Chapter.project_id == project_id)
            .order_by(desc(Chapter.chapter_no), desc(ChapterSummary.id))
            .limit(safe_limit)
            .all()
        )

    return [
        {
            "id": row.id,
            "chapter_id": row.chapter_id,
            "chapter_no": row.chapter_no,
            "title": row.title,
            "summary": row.summary,
            "character_changes": row.character_changes,
            "world_changes": row.world_changes,
            "new_foreshadowings": row.new_foreshadowings,
            "timeline_events": row.timeline_events,
            "source_run_id": row.source_run_id,
            "source_version_id": row.source_version_id,
            "source_version_number": row.source_version_number,
            "created_at": row.created_at.isoformat() if row.created_at else None,
        }
        for row in rows
    ]


@router.get("/{project_id}/analysis-status")
def chapter_analysis_status(project_id: int) -> list[dict]:
    """读取章节记忆补充状态，供批量回填和进度统计使用。

    步骤 1：按项目读取章节基本信息和正文长度，不传输整章正文。
    步骤 2：关联摘要记录，标记章节是否已经完成过分析。
    步骤 3：按章节号排序，返回前端可直接用于续跑的状态清单。
    """
    with get_business_db() as db:
        summary_query = (
            db.query(ChapterSummary.chapter_id.label("chapter_id"))
            .filter(
                ChapterSummary.summary.isnot(None),
                ChapterSummary.summary != "",
                ~ChapterSummary.summary.like("开发模式摘要%"),
            )
            .distinct()
            .subquery()
        )
        rows = (
            db.query(
                Chapter.id,
                Chapter.chapter_no,
                Chapter.title,
                Chapter.status,
                func.coalesce(func.length(Chapter.content), 0).label("content_length"),
                summary_query.c.chapter_id.label("analyzed_chapter_id"),
            )
            .outerjoin(summary_query, summary_query.c.chapter_id == Chapter.id)
            .filter(Chapter.project_id == project_id)
            .order_by(Chapter.chapter_no.asc(), Chapter.id.asc())
            .all()
        )

    return [
        {
            "chapter_id": row.id,
            "chapter_no": row.chapter_no,
            "title": row.title,
            "status": row.status,
            "content_length": row.content_length or 0,
            "has_content": (row.content_length or 0) > 0,
            "has_summary": row.analyzed_chapter_id is not None,
        }
        for row in rows
    ]


@router.get("/{project_id}/context-preview")
def context_preview(
    project_id: int,
    chapter_no: int = 1,
    outline_id: int | None = None,
    query: str = "",
    selection: str = "",
) -> dict:
    """预览章节生成会读取的上下文包。

    步骤 1：解析前端提交的上下文选择。
    步骤 2：将补充要求和选择项交给正式上下文检索器。
    """
    # 步骤 1：验证选择参数为 JSON 对象，避免预览请求静默退回自动推荐。
    try:
        context_selection = json.loads(selection) if selection else None
    except json.JSONDecodeError as exc:
        raise HTTPException(status_code=422, detail="上下文选择格式无效") from exc
    if context_selection is not None and not isinstance(context_selection, dict):
        raise HTTPException(status_code=422, detail="上下文选择必须是对象")

    # 步骤 2：把用户要求和手动选择一并交给实际生成使用的检索器。
    return build_context_preview(project_id, chapter_no, outline_id, query, context_selection)


@router.post("/chapter-draft")
def chapter_draft(payload: ChapterDraftRequest) -> dict:
    """生成章节正文。

    支持指定大纲，也支持覆盖已有章节草稿。
    """
    return draft_chapter(
        payload.project_id,
        payload.chapter_no,
        payload.instruction,
        payload.rhythm_level,
        payload.outline_id,
        payload.chapter_id,
        payload.context_selection,
    )


@router.post("/chapter-draft/stream")
def chapter_draft_stream(payload: ChapterDraftRequest) -> StreamingResponse:
    """流式生成章节正文。

    使用 NDJSON：每行一个事件对象，前端可以边读边追加到正文编辑区。
    """

    def event_lines() -> Iterator[str]:
        try:
            for event in draft_chapter_stream(
                payload.project_id,
                payload.chapter_no,
                payload.instruction,
                payload.rhythm_level,
                payload.outline_id,
                payload.chapter_id,
                payload.context_selection,
            ):
                yield json.dumps(event, ensure_ascii=False) + "\n"
        except Exception as exc:  # noqa: BLE001
            # 流式响应头一旦发出，后续异常不能再变成标准 500；用 error 事件交给前端做可恢复提示。
            yield json.dumps(
                {
                    "type": "error",
                    "message": f"流式生成中断：{exc}",
                    "trace": traceback.format_exc(limit=2),
                },
                ensure_ascii=False,
            ) + "\n"

    return StreamingResponse(event_lines(), media_type="application/x-ndjson")


@router.post("/chapter-edit/chat")
def chapter_edit_chat(payload: ChapterEditChatRequest) -> dict:
    """讨论或提出章节改稿候选，绝不在此接口直接覆盖正文。"""
    with get_business_db() as db:
        chapter = db.query(Chapter.id).filter(
            Chapter.id == payload.chapter_id,
            Chapter.project_id == payload.project_id,
        ).first()
    if not chapter:
        raise HTTPException(status_code=404, detail="章节不存在或不属于当前项目")
    try:
        return propose_chapter_edit(
            project_id=payload.project_id,
            chapter_no=payload.chapter_no,
            chapter_title=payload.chapter_title,
            outline_id=payload.outline_id,
            content=payload.content,
            scope=payload.scope,
            selection_start=payload.selection_start,
            selection_end=payload.selection_end,
            instruction=payload.instruction,
            conversation=payload.conversation,
            context_selection=payload.context_selection,
        )
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    except LLMError as exc:
        raise HTTPException(status_code=503, detail=str(exc)) from exc


@router.post("/chapter-edit/apply")
def chapter_edit_apply(payload: ChapterEditApplyRequest) -> dict:
    """校验原稿仍未变化后，保存候选正文并创建新的当前版本。"""
    result = WorkflowPersistence.apply_chapter_edit(
        project_id=payload.project_id,
        chapter_id=payload.chapter_id,
        expected_content=payload.expected_content,
        revised_content=payload.revised_content,
        summary=payload.summary,
    )
    if result is None:
        raise HTTPException(
            status_code=409,
            detail="章节正文已发生变化。请刷新当前内容后重新生成改稿候选，避免覆盖新修改。",
        )
    return {"success": True, **result}


@router.post("/chapter-analyze")
def chapter_analyze(payload: ChapterAnalyzeRequest) -> dict:
    """分析章节正文并沉淀摘要。"""
    try:
        return analyze_chapter(payload.project_id, payload.chapter_id, payload.content)
    except LookupError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc


@router.post("/polish")
def polish(payload: PolishRequest) -> dict:
    """按指定模式精修章节正文。"""
    return polish_chapter(payload.project_id, payload.chapter_id, payload.mode, payload.instruction)


@router.post("/consistency-check")
def consistency_check(payload: ConsistencyCheckRequest) -> dict:
    """检查章节与资料库是否存在明显缺口或冲突。"""
    return check_consistency(payload.project_id, payload.chapter_id, payload.content)


@router.post("/volume-analyze")
def volume_analyze(payload: VolumeAnalyzeRequest) -> dict:
    """分析卷设定并自动更新大纲总览。

    读取卷的描述、核心事件、章节规划等信息，通过 LLM 分析后
    补充完善大纲总览（主线、核心冲突、结局走向等），
    为后续章节生成 Agent 提供更清晰的创作方向。
    """
    return analyze_volume(payload.project_id, payload.volume_id, payload.instruction)

File: frontend/src/api/agents.ts
import { apiClient } from './client'
import type { ChapterDraftResult, ChapterSummary, ConsistencyCheckResult, ContextPreview, GenerationLog } from '@/types/domain'

export type ChapterDraftStreamEvent =
  | { type: 'start'; message: string; chapter_id?: number }
  | { type: 'delta'; content: string }
  | { type: 'done'; chapter_id: number; title: string; source: string }
  | { type: 'error'; message: string; trace?: string }

export interface ChapterDraftStreamHandlers {
  onStart?: (message: string, chapterId?: number) => void
  onDelta?: (content: string) => void
  onDone?: (result: Omit<ChapterDraftResult, 'content'>) => void
  onError?: (message: string) => void
}

export async function getAgentLogs(projectId: number, limit = 20) {
  // Agent 日志用于工作台右侧时间线，只读取最近记录，避免前端一次拉取过多历史。
  const { data } = await apiClient.get<GenerationLog[]>(`/agents/${projectId}/logs`, { params: { limit } })
  return data
}

export async function getChapterSummaries(projectId: number, limit = 20) {
  // 长期记忆第一版：先读取结构化章节摘要列表，未来可替换为检索排序结果。
  const { data } = await apiClient.get<ChapterSummary[]>(`/agents/${projectId}/summaries`, { params: { limit } })
  return data
}

export interface NextChapterTarget {
  available: boolean
  reason: string
  outline_id: number | null
  chapter_no: number | null
  outline_title: string
  instruction: string
  chapter_id: number | null
  chapter_title: string
  position: number | null
  total_outlines: number
}

export async function getNextChapterTarget(projectId: number) {
  // 目标由后端按当前已保存的大纲顺序解析，避免页面缓存或拖动排序后生成错章。
  const { data } = await apiClient.get<NextChapterTarget>(`/agents/${projectId}/next-chapter`)
  return data
}

export interface ChapterAnalysisStatus {
  chapter_id: number
  chapter_no: number
  title: string
  status: string
  content_length: number
  has_content: boolean
  has_summary: boolean
}

export async function getChapterAnalysisStatus(projectId: number) {
  // 只读取章节记忆回填状态和正文长度，避免为统计进度重复传输整章正文。
  const { data } = await apiClient.get<ChapterAnalysisStatus[]>(`/agents/${projectId}/analysis-status`)
  return data
}

export async function getContextPreview(
  projectId: number,
  chapterNo: number,
  outlineId?: number | null,
  query = '',
  selection?: Record<string, number[]>,
) {
  // 上下文包预览只读不写，用于生成前确认 Agent 实际会读取哪些资料。
  const { data } = await apiClient.get<ContextPreview>(`/agents/${projectId}/context-preview`, {
    params: {
      chapter_no: chapterNo,
      outline_id: outlineId ?? undefined,
      query: query || undefined,
      selection: selection ? JSON.stringify(selection) : undefined,
    }
  })
  return data
}

export async function draftChapter(payload: Record<string, unknown>) {
  // 章节生成：后端会读取大纲、世界观、角色、组织、伏笔等上下文。
  const { data } = await apiClient.post<ChapterDraftResult>('/agents/chapter-draft', payload)
  return data
}

export async function draftChapterStream(
  payload: Record<string, unknown>,
  handlers: ChapterDraftStreamHandlers
): Promise<Omit<ChapterDraftResult, 'content'> | null> {
  // 流式章节生成使用 fetch 读取 NDJSON；Axios 在浏览器侧不适合逐块消费响应体。
  const controller = new AbortController()
  const timeoutId = window.setTimeout(() => controller.abort(), 300000)
  const response = await fetch('/backend-api/agents/chapter-draft/stream', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
    signal: controller.signal
  })
  if (!response.ok || !response.body) {
    window.clearTimeout(timeoutId)
    throw new Error(`流式生成请求失败：${response.status}`)
  }

  const reader = response.body.getReader()
  const decoder = new TextDecoder()
  let buffer = ''
  let doneResult: Omit<ChapterDraftResult, 'content'> | null = null

  function consumeLine(line: string) {
    if (!line.trim()) return
    const event = JSON.parse(line) as ChapterDraftStreamEvent
    if (event.type === 'start') handlers.onStart?.(event.message, event.chapter_id)
    if (event.type === 'delta') handlers.onDelta?.(event.content)
    if (event.type === 'error') {
      handlers.onError?.(event.message)
      throw new Error(event.message)
    }
    if (event.type === 'done') {
      doneResult = {
        chapter_id: event.chapter_id,
        title: event.title,
        source: event.source
      }
      handlers.onDone?.(doneResult)
    }
  }

  try {
    while (true) {
      const { value, done } = await reader.read()
      if (done) break
      buffer += decoder.decode(value, { stream: true })

      const lines = buffer.split('\n')
      buffer = lines.pop() ?? ''
      for (const line of lines) {
        consumeLine(line)
      }
    }

    buffer += decoder.decode()
    consumeLine(buffer)
  } catch (error) {
    if (error instanceof DOMException && error.name === 'AbortError') {
      throw new Error('流式生成超时：5 分钟内未完成')
    }
    throw error
  } finally {
    window.clearTimeout(timeoutId)
    reader.releaseLock()
  }

  return doneResult
}

export async function analyzeChapter(payload: Record<string, unknown>) {
  // 章节分析：把正文拆成摘要、人物变化、世界观变化等长期记忆。
  const { data } = await apiClient.post<{
    chapter_id: number
    analysis: string
    summary: string
    character_changes: string
    world_changes: string
    new_foreshadowings: string
    timeline_events: string
    pending_change_count?: number
    analysis_status?: 'complete' | 'unavailable'
  }>('/agents/chapter-analyze', payload)
  return data
}

export async function polishChapter(payload: Record<string, unknown>) {
  // 章节精修：保留事实基础上按指定模式重写当前章节。
  const { data } = await apiClient.post<{ chapter_id: number; content: string }>('/agents/polish', payload)
  return data
}

export interface ChapterEditChatResult {
  action: 'discussion' | 'proposal'
  reply: string
  candidate_text: string
  scope: 'chapter' | 'selection'
  selection_start: number | null
  selection_end: number | null
  usage: { input_tokens?: number; output_tokens?: number; total_tokens?: number } | null
}

export async function chatChapterEdit(payload: Record<string, unknown>) {
  // 对话改稿只返回讨论答复或候选正文；确认应用另走版本化保存接口。
  const { data } = await apiClient.post<ChapterEditChatResult>('/agents/chapter-edit/chat', payload)
  return data
}

export async function applyChapterEdit(payload: Record<string, unknown>) {
  // 确认候选时一并提交原稿快照，服务端会拒绝覆盖期间发生的其他编辑。
  const { data } = await apiClient.post<{
    success: boolean
    chapter_id: number
    version_id: string
    version_number: number
    word_count: number
  }>('/agents/chapter-edit/apply', payload)
  return data
}

export async function checkConsistency(payload: Record<string, unknown>) {
  // 一致性检查：当前支持 fallback 结构化结果，后续可无缝接入真实模型判断。
  const { data } = await apiClient.post<ConsistencyCheckResult>('/agents/consistency-check', payload)
  return data
}

export async function analyzeVolume(payload: Record<string, unknown>) {
  // 卷分析：分析卷设定并自动更新大纲总览，为章节生成 Agent 提供创作方向。
  const { data } = await apiClient.post<{
    volume_id: number
    overview_id: number
    analysis: string
    main_plot: string
    core_conflict: string
    ending: string
    volume_summary: string
    chapter_suggestions: string
    source: string
  }>('/agents/volume-analyze', payload)
  return data
}
